### ISSUE TITLE
Generally enable reading secrets from files

### ISSUE THREAD

**Issue by beorn7** [prometheus/prometheus#8551]
For a bunch of secrets in the config file, we already allow to alternatively read it from a file, e.g. if there is a `password` field, there is also a `password_file` field. Only one of those may be used.

However, we do not offer this behavior consistently. Occasionally, we get feature requests to allow it for particular fields. This issue proposes to just consistently offer the feature for all secrets in general (exceptions might be possible or even required in particular cases, but those would then be exceptions and not just the normal case).

See also [this discussion](https://groups.google.com/g/prometheus-developers/c/10duXKRNOuc/m/PZnyLy6xBAAJ).

Note that the semantics generally is to read the secret from the file for each new request (so that password rotation works seamlessly). Exceptions need to be documented clearly (but ideally, we would avoid those entirely). Cf. a previous issue about this: #6140

Also note that for consistency with the already quite frequent existing usage, we won't go for a new YAML construct like 
```Yaml
secret:
  type: file
  file: "/etc/my-secrets/..."
```

This could be considered for a future release, though.

**Comment by beorn7** [comment]
@mem wants to work on this, at least on a few secrets, so assigning to him.

**Comment by mem** [comment]
For prometheus in particular, these are the config entries specifying file paths to obtain secrets.

These are part of [TLSConfig](https://github.com/prometheus/common/blob/47ee79a16821e03c31ca741659d4f7919aed91c1/config/http_config.go#L436):

* CAFile, from github.com/prometheus/common/config/http_config.go
* CertFile, from github.com/prometheus/common/config/http_config.go
* KeyFile, from github.com/prometheus/common/config/http_config.go

This is part of [HTTPClientConfig](https://github.com/prometheus/common/blob/47ee79a16821e03c31ca741659d4f7919aed91c1/config/http_config.go#L104):

* BearerTokenFile, from github.com/prometheus/common/config/http_config.go, this is already handled in the desired way,  as there's another entry BearerToken

This is part of BasicAuth:

* PasswordFile, from github.com/prometheus/common/config/http_config.go, this is already handled in the desired way with the entry Password.

This is part of [SDConfig](https://github.com/prometheus/prometheus/blob/main/discovery/marathon/marathon.go#L73)

* AuthTokenFile, already handled in the desired way with AuthToken.

**Comment by roidelapluie** [comment]
￼ CAFile, from github.com/prometheus/common/config/http_config.go
￼ CertFile, from github.com/prometheus/common/config/http_config.go
￼ KeyFile, from github.com/prometheus/common/config/http_config.go

Are already files


￼ TLSCertPath
￼ TLSKeyPath
￼ ClientCAs


Are already files

**Comment by roidelapluie** [comment]
I think the main goal here is to be able to read secrets from external files, not to load certificates from primary file.

**Comment by mem** [comment]
The agreement in the linked discussion is to make this consistent, so those options which only offer the possibility of reading a file should have the inline form added.

I was actually looking for the options where the user can only specify inline secrets, and I haven't found any, but I have only looked at the mentioned packages.

**Comment by roidelapluie** [comment]
here are the secrets I think we can't read from files:

- azure sd's client secret
- consul token
- aws secret key
- openstack password
- openstak application_credential_secret

**Comment by mem** [comment]
ah, thanks!

**Comment by mem** [comment]
Prometheus
----------------

- [ ] [Azure SD's client secret](https://github.com/prometheus/prometheus/blob/main/discovery/azure/azure.go#L79)
- [ ] [Consul SD's token](https://github.com/prometheus/prometheus/blob/main/discovery/consul/consul.go#L112)
- [ ] [Consul SD's password](https://github.com/prometheus/prometheus/blob/main/discovery/consul/consul.go#L117)
- [ ] [EC2 SD's secret key](https://github.com/prometheus/prometheus/blob/main/discovery/ec2/ec2.go#L85)
- [x] [Marathon SD's auth token](https://github.com/prometheus/prometheus/blob/main/discovery/marathon/marathon.go#L76)
- [ ] [OpenStack SD's password](https://github.com/prometheus/prometheus/blob/main/discovery/openstack/openstack.go#L51)
- [ ] [OpenStack SD's application credential secret](https://github.com/prometheus/prometheus/blob/main/discovery/openstack/openstack.go#L58)

**Comment by LeviHarrison** [comment]
Pretty sure #8926 does both the token and password for Consul.

**Comment by roidelapluie** [comment]
Yes

**Comment by LeviHarrison** [comment]
If @mem isn't working on this anymore I can pick it up.

**Comment by mem** [comment]
Sorry about the radio silence, I'm picking this up again. Thanks for the reminder!

**Comment by roidelapluie** [comment]
I do not think that we should 

> I think the main goal here is to be able to read secrets from external files, not to load certificates from primary file.

I stand by this. Certificates are files almost everywhere, it is not natural to load them inline.

**Comment by beorn7** [comment]
Note: The discussion is now also happening in https://github.com/prometheus/common/pull/318 , which is @mem's implementation of what's suggested here for TLS.

My thoughts: I totally get @mem's argument about consistency both ways, but I have to admit, when I emphasized consistency in the discussion on the mailing list, I was only thinking about fields with inline secrets so far, i.e. that we should rather give all of them the option to provide the secret in a file than just selected few. I honestly had not thought about secrets that we used to _only_ provide as a file and never inline.

Seeing the implementation in https://github.com/prometheus/common/pull/318 and @roidelapluie's argument that is very uncommon to provide TLS certificates inline and that users never requested the latter.

So perhaps we can formulate the consistency somewhat weaker: We consistently want to enable providing secrets in files. However, there is no requirement for an option to represent a secret inline in the config file.
The latter comes in handy for short secrets in text format (like passwords), and it's an encouraged feature in those cases, but it's rarely desired to put long and/or binary-encoded secrets (like TLS certificates) inline into the config file.

**Comment by mem** [comment]
OK, I can understand that the case for inline CA certificates is weak.

For client keys on the other hand the use case is a large blackbox_exporter configuration where the secrets are stored in a central location and the BBE configuration is generated, not written by hand, so it's actually simpler to manage a single configuration file instead of a configuration file plus a number of different certificates and their keys. It's neither more or less secure (if the configuration file with the embedded secrets is in the filesystem, that is just as secure or insecure has having the configuration file plus the multiple certificates in the filesystem), it's just simpler. I can understand the simplicity of having a file be reloaded over and over again in case it ever changes, every time a scrape job runs, 

(I mention blackbox_exporter because it's the example where it's most obvious to me how this can be used, but the same thing applies to Prometheus with scrape jobs that require different client authentication secrets, which might sound artificial, but it was actually the case at `$PAST_JOB` in a slightly different context)

I'll make the other changes I mentioned before.

**Comment by beorn7** [comment]
I guess if there is an actual need to inline keys (or even certs), then we can just offer it. My impression was that @roidelapluie was opposed to offering the feature merely for consistency reasons without anyone having ever asked for it.

**Comment by roidelapluie** [comment]
Yes I am opposed to it. Secrets are multi line and just the standard is to keep them away in separated files. Multi line in YAML is something I really want to avoid.

**Comment by beorn7** [comment]
Then let's just separate things and really just enable reading secrets from files as part of this issue.

Inlining secrets that can only be read from files right now is then a separate topic to discuss later.

**Comment by fsdrw08** [comment]
any updates? I would like to specify consul sd token from file

**Comment by roidelapluie** [comment]
> any updates? I would like to specify consul sd token from file

This works:

```yaml
scrape_configs:
- job_name: consul
 consul_sd_configs:
 - server: localhost:8500
   authorization:
     credentials_file: consul_token
```
